draft-ietf-taps-interface.md

```diff
@@ -1185,30 +1193,41 @@ number of Messages with a single call. This allows an application to provide
 backpressure to the transport stack when it is temporarily not ready to
 receive messages.
 
+Receive takes an optional minIncompleteLength argument. This value indicates
+the smallest partial Message data size that should be delivered. By default, this
+value is infinite, which means that only complete, atomic Messages will be delivered.
+If this value is set to some smaller value, the associated ReceiveHandler will
+be triggered only when at least that many bytes are available, the Message is
+complete, or the system needs to free up memory.
+
 Receive also takes an optional maxLength argument, the maximum size (in bytes
 of data) Message the application is currently prepared to receive. The default
 value for maxLength is infinite. If an incoming Message is larger than the
 minimum of this size and the maximum Message size on receive for
 the Connection's Protocol Stack, it will be received as a partial Message.
-Note that maxLength does not guarantee that the application will receive that
-many bytes if they are available; the interface may return partial Messages
-smaller than maxLength according to implementation constraints.
+Partial Messages are indicated by marking the endOfMessage boolean value
+as false in the ReceivedHandler. Note that maxLength does not guarantee that 
+the application will receive that many bytes if they are available; the interface may
+return partial Messages smaller than maxLength according to implementation constraints.
 
 ~~~
-Connection -> Received<Message>
+Connection -> Received<Message, messageData, endOfMessage>
 ~~~
 
 As with sending, the type of the Message to be passed is dependent on the
 implementation, and on the constraints on the Protocol Stacks implied by the
 Connection's transport parameters. 
 
-The Message Object must provide some method to retrieve an octet array
-containing application data, corresponding to a single message within the
-underlying Protocol Stack's framing.  See {{receive-framing}} for handling
-framing in situations where the Protocol Stack provides octet-stream transport
-only.
+The Message Object identifies the logical message received by the transport
+protocol, and provides methods to access metadata about the received data.
+The octets of data associated with this message are delivered as messageData.
+Multiple invocations of the ReceivedHandler may devlier data for the same
+Message until the endOfMessage flag is delivered. See {{receive-framing}} for 
```

Comment:
Partial message data is sent in-order. Received data may be received out-of-order, right? Do we need to clarify that?

Reply:
This is THE problem I have with this whole thing - it's also the question that I raised at the end of #200 and don't see addressed. Partial messages may arrive out of order, or they may never arrive at all, if we map them on an unreliable transport. At least, applications must have a number to identify that they're now getting message fragment 3 out of 7, something like that.

Reply:
And what about partial reliability?  E.g., fragments 1 and 2 out of a total of 3 arrive in time, but for the third, the time expires and the transport gives up. Shouldn't it then tell the receiving application that it will never get the third fragment, and should throw away the first two?

Reply:
My inclination here is to only receive chunks in order and not deliver gaps. Does that work for you?

Reply:
@tfpauly I think that's a good direction, as it makes things much easier. I think we should then still add a way for the transport system to signal to the application: "throw away your chunks, the rest is never going to arrive" - e.g. to support partial reliability.

But I have a bigger concern, then: what about UDP? What if I send a large message, the chosen transport is UDP, and the message is split it into 3 packets - first, this is perhaps not a great idea; second, the receiver won't be able to know about the order, so can't even do what you suggest here. How are you going to handle this case?

Reply:
So, in our implementation, we allow the protocol to require that data for a single message fits within a limit (for UDP, whatever it can fit in a datagram). Splitting up a message over multiple UDP datagrams without any other framing layer on top doesn't make any sense, since there isn't any way for the other side to understand the message. So, in our implementation, we throw an error if you try to do this.

Protocols on top of UDP, like IKE, may support fragmenting large messages across datagrams, but then the message that you send is first interpreted by a protocol in the stack that supports unbounded message sizes.

I also do agree that there should be a signal that the chunks won't ever arrive—our implementation passes an optional error along with the ReceiveHandler to indicate this.

Reply:
Ah ok!  Great, I think that solves it. I'll conclude by requesting the following text:
1) in the API draft, some kind of a signal that the chuncks won't error arrive (as we agreed)
2) in the API draft, for the receiver side, a hint that message chunks will only arrive in order
3) in the implementation draft, a hint that message fragmentation of course requires support by the underlying protocol and thus the size of any message, even with fragmentation, will have to be limited to the message size that the underlying protocol supports. For example, messages can not be spread across multiple UDP packets because the receiver would not understand that the message parts belong together.

I can take care of 2 and 3 (but I'm fine with you doing it too if you want!  Just offering to help instead of always only criticizing!) .... but for 1), I think it would be better for you to do it since your implementation already does it.  Ok?